### PR TITLE
[RSDK-9981] - implement blurry classifier

### PR DESCRIPTION
**Pull request by jgeaso1266**
## Future Actions
- update confidence to use linear interpolation

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -10,3 +10,8 @@ setup: .installed
 .PHONY: dist/main
 dist/archive.tar.gz: .installed
 	./build.sh
+
+test:
+	PYTHONPATH=./src pytest
+lint:
+	pylint --disable=C0301,C0114,W0401,C0115,R0913,R0917,E1101,W0614 src/
```
Comment (penguinland) [inline Makefile:17]: What are all these rules that you're disabling? Why disable them globally? Ranked from most preferred to least preferred,
- Don't disable anything. Follow all the linting rules
- Disable certain rules locally, with a comment about why we're disabling the rule on this one line
- Disable certain rules globally (on this line) with a comment about what the rule is for and why we expect to never follow it.

See https://github.com/viam-modules/motion-detector/pull/39 as an example of what I'd prefer. (unrelated: I'm waiting on a review from you on that PR!)

File: requirements.txt [hunk 2]
```diff
@@ -1,3 +1,10 @@
 pyinstaller==6.12.0
 typing-extensions==4.12.2
 viam-sdk==0.41.0
+opencv-python
+pillow
+
+pytest
+pytest-asyncio
+pylint
```
Comment (penguinland) [inline requirements.txt:9]: Two requests. 
1. Pin a specific version of each of these. Otherwise, a year from now some new employee is going to install these requirements and the code will work on your machine and fail on the new employee's machine because these dependencies had a breaking change between now and then. 
2. Alphabetize this file, so it's easier to check if a certain dependency is already in the list.

Both of these can be done by running `pip freeze` and putting its output in this file. (that approach will also include all the indirect dependencies, which I don't really care about one way or the other)

**Review by penguinland — COMMENTED** [review]
I haven't looked at the "real" code yet, but here's feedback around the edges

File: Makefile (hunk 1 above)
Comment (jgeaso1266) [reply Makefile:17]: Very good point. Will make the edit.

File: tests/test_blurry_classifier.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            "camera_name": CAMERA_NAME,
+            "blurry_threshold": BLURRINESS_THRESHOLD,
+        }
+    ))
+
+def get_vision_service():
+    service = BlurryDetector("test")
+    cam = FakeCamera(CAMERA_NAME)
+    camera_name = cam.get_resource_name(CAMERA_NAME)
+    cfg = config
```
Comment (penguinland) [inline tests/test_blurry_classifier.py:27]: What's the point of this line? If you're trying to make a deep copy of the original, you haven't. Can we just use `config` in the next 2 lines and get rid of this variable?

File: tests/test_blurry_classifier.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    return service
+
+class TestBlurryDetector:
+    def test_empty(self):
+        blurry_detector = BlurryDetector("test_blurry_detector")
+        with pytest.raises(ValueError, match="A camera name is required for face_identification vision service module."):
+            blurry_detector.validate_config(ServiceConfig(attributes=dict_to_struct({})))
+
+    @pytest.mark.asyncio
+    @patch('viam.components.camera.Camera.get_resource_name', return_value="fake_cam")
```
Comment (penguinland) [inline tests/test_blurry_classifier.py:39]: Is `return_value` supposed to be `CAMERA_NAME`? If so, use that here. If not, put in a comment explaining what's different. Same comment on other tests, too. (a single comment explaining all of them is sufficient if you go that route)

File: tests/test_blurry_classifier.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        assert blurry_detector.camera is not None
+
+    @pytest.mark.asyncio
+    @patch('viam.components.camera.Camera.get_resource_name', return_value="fake_cam")
+    async def test_get_classifications(self, fake_cam):
+        blurry_detector = get_vision_service()
+
+        # Test with a blurry image
+        result = await blurry_detector.get_classifications(
+            image= await blurry_detector.camera.get_image(True),
```
Comment (penguinland) [inline tests/test_blurry_classifier.py:54]: Mild preference to mention the argument by name. I suspect this is calling `.get_image(mime_type=True)`, and I suspect you meant `.get_image(blurry=True)`. Even if I've made a mistake and the code works correctly as-is, I'd still specify the argument name so that people like me can avoid the mistake.

File: tests/test_blurry_classifier.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        # Test with a non-blurry image
+        result = await blurry_detector.get_classifications(
+            image=await blurry_detector.camera.get_image(blurry=False),
+            count=1,
+        )
+
+        assert result == []
+
+    @pytest.mark.asyncio
+    @patch('viam.components.camera.Camera.get_resource_name', return_value="fake_cam")
```
Comment (penguinland) [inline tests/test_blurry_classifier.py:70]: This is super minor and nitpicky, but why do you sometimes use single-quotes for strings, and sometimes double-quotes? I try to pick only one of them and use that consistently throughout, so other people don't need to wonder if I'm trying to convey additional meaning by using one or the other.

File: tests/fake_camera.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+from PIL import Image
+from viam.components.camera import Camera
+from viam.gen.component.camera.v1.camera_pb2 import GetPropertiesResponse
+from viam.media.utils import pil
+from viam.media.video import CameraMimeType, NamedImage, ViamImage
+from viam.proto.common import ResponseMetadata
+
+
+def read_image(name: str):
+    return Image.open(os.path.join("./tests", "img", name + ".jpg"))
```
Comment (penguinland) [inline tests/fake_camera.py:13]: This is probably good enough as-is, but will break if you run the tests from anywhere besides the root of the repo.

If you want to fix it, try replacing `"./tests"` with `os.path.dirname(__file__)`. If you don't want to fix it, I'm fine with leaving it as-is, too.

File: src/utils.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+                "Unsupported image type: %s. Supported types are %s.",
+                image.mime_type,
+                SUPPORTED_IMAGE_TYPE,
+            )
+
+            raise ValueError(f"Unsupported image type: {image.mime_type}.")
+
+        im = Image.open(BytesIO(image.data), formats=LIBRARY_SUPPORTED_FORMATS).convert(
+            "RGB"
+        )  # convert in RGB png openened in RGBA
```
Comment (penguinland) [inline src/utils.py:39]: typo: openened

More broadly, I'm not really sure what this comment means. Is it "remove any transparency"?

File: src/utils.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        return np.array(im)
```
Comment (penguinland) [inline src/utils.py:40]: The last few lines of this if statement are identical to the last few lines of the function. I'd refactor slightly:
```python
if isinstance(image, ViamImage):
    ...
    pil_image = Image.open(BytesIO(image.data), formats=...)
else:
    pil_image = image
result = pil_image.convert("RGB")
return np.array(result)
```

**Review by penguinland — COMMENTED** [review]
I've got 1 more file to go, but it's the main file. 

I worry that some of my feedback is too nitpicky. You're welcome to say "I hear your point, Alan, but I'm not going to do it." These are suggestions, not commands.

File: src/blurry_classifier_module.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+        Returns:
+            Self: The resource
+        """
+        service = cls(config.name)
+        service.reconfigure(config, dependencies)
+        return service
+
+    @classmethod
+    def validate_config(cls, config: ComponentConfig) -> Sequence[str]:
```
Comment (penguinland) [inline src/blurry_classifier_module.py:67]: A note for myself: it feels like line 66 should be `@staticmethod` and this line should have the `cls` argument removed. That gets auto-added if you generate your skeleton from the CLI, as a way to demonstrate to people that this kind of thing is possible. It's not worth the hassle to change every time; just leave it be.

File: Makefile [hunk 11]
```diff
@@ -10,3 +10,8 @@ setup: .installed
 .PHONY: dist/main
 dist/archive.tar.gz: .installed
 	./build.sh
+
+test:
+	PYTHONPATH=./src pytest
+lint:
+	pylint --disable=C0114,C0115,E1101 src/
```
Comment (penguinland) [inline Makefile:17]: I think E1101 is about creating members in a function that is not `__init__`. If we stopped disabling that, which lines would fail the lint check? I don't see any, but maybe I missed something.

File: src/blurry_classifier_module.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+        return_detections: bool = False,
+        return_object_point_clouds: bool = False,
+        *,
+        extra: Optional[Mapping[str, ValueTypes]] = None,
+        timeout: Optional[float] = None,
+    ) -> CaptureAllResult:
+        if camera_name not in (self.camera_name, ""):
+            raise ValueError(
+                "Camera name ",
+                {camera_name},
```
Comment (penguinland) [inline src/blurry_classifier_module.py:120]: Why put size-1 sets into the argument list like this? I suspect you meant to make this a format string, rather than giving the `ValueError` constructor so many arguments.
```python
raise ValueError(f"Camera name {camera_name} does not match the name {self.camera_name} in the config")
```

File: src/blurry_classifier_module.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        image: ViamImage,
+        count: int,
+        *,
+        extra: Optional[Mapping[str, ValueTypes]] = None,
+        timeout: Optional[float] = None,
+    ) -> List[Classification]:
+        img = decode_image(image)
+        blurriness = cv.Laplacian(img, cv.CV_64F).var()
+        if blurriness < self.blurry_threshold:
+            return [Classification(class_name="blurry", confidence=0.5)]
```
Comment (penguinland) [inline src/blurry_classifier_module.py:182]: Not a fan of hardcoding the confidence like this. Particularly a value of 0.5, suggesting we're no more confident than a coin flip.

We want the following properties:
- if the blurriness is right on the threshold, the confidence is 0.5 to indicate our uncertainty
- the minimum blurriness is 0, if the image is a flat monochrome
- The maximum blurriness is 1040400, if the image is a checkerboard of white and black pixels.
- So, we need a way to map blurriness into the [0, 1] range such that 0 maps to 0, 1040400 maps to 1, and the threshold maps to 0.5. Pure linear interpolation is a bad idea: in practice we'll never see blurriness as high as the theoretical max.
- Maybe we do linear interpolation in the range [0, 2*threshold], and then just 1.0 for any value above that? We could also try some activation function from neural nets (sigmoid? hyperbolic tangent?), though that feels like overkill if we're not going to need gradient descent to train a model.

File: src/blurry_classifier_module.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    async def get_classifications(
+        self,
+        image: ViamImage,
+        count: int,
+        *,
+        extra: Optional[Mapping[str, ValueTypes]] = None,
+        timeout: Optional[float] = None,
+    ) -> List[Classification]:
+        img = decode_image(image)
+        blurriness = cv.Laplacian(img, cv.CV_64F).var()
```
Comment (penguinland) [inline src/blurry_classifier_module.py:180]: nitpick: this variable feels backwards. I'd expect high values to indicate _non_ blurriness. but I don't have a good word for that. :shrug: Maybe it's fine as-is.

File: src/blurry_classifier_module.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+    ) -> Vision.Properties:
+        return Vision.Properties(
+            classifications_supported=True,
+            detections_supported=False,
+            object_point_clouds_supported=False,
+        )
+
+
+if __name__ == "__main__":
+    asyncio.run(Module.run_from_registry())
```
Comment (penguinland) [inline src/blurry_classifier_module.py:210]: I think these 2 lines should be removed: I don't think there is ever a reason for this file to be the main executable.

File: src/blurry_classifier_module.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+        timeout: Optional[float] = None,
+    ) -> List[Classification]:
+        img = decode_image(image)
+        blurriness = cv.Laplacian(img, cv.CV_64F).var()
+        if blurriness < self.blurry_threshold:
+            return [Classification(class_name="blurry", confidence=0.5)]
+
+        return []
+
+
```
Comment (penguinland) [inline src/blurry_classifier_module.py:186]: Sometimes you're separating functions with 1 blank line, sometimes 2. Be consistent. IIRC PEP8 says 1 blank line in between methods within a class, and 2 blank lines in between top-level things like class definitions.

File: src/blurry_classifier_module.py [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+        self,
+        image: ViamImage,
+        count: int,
+        *,
+        extra: Optional[Mapping[str, ValueTypes]] = None,
+        timeout: Optional[float] = None,
+    ) -> List[Classification]:
+        img = decode_image(image)
+        blurriness = cv.Laplacian(img, cv.CV_64F).var()
+        if blurriness < self.blurry_threshold:
```
Comment (penguinland) [inline src/blurry_classifier_module.py:181]: I'd put in a logger and have debug logs of what the blurriness value is. That way, I have an easier time setting the right threshold when I want to use this.

**Review by penguinland — COMMENTED** [review]
Most of this feedback is minor nitpicks and you can ignore anything you don't like. but the part about what confidence we return is important.

and changing the confidence returned might change the behavior of the tests. I'd just assert that the confidence returned is in the range of values we expect it to be near.

File: tests/test_blurry_classifier.py (hunk 3 above)
Comment (jgeaso1266) [reply tests/test_blurry_classifier.py:27]: Good catch. I was considering making a deep copy and changing some attributes of `config`, but decided against it and still kep tthe `cfg` variable.

File: tests/test_blurry_classifier.py (hunk 4 above)
Comment (jgeaso1266) [reply tests/test_blurry_classifier.py:39]: Agreed. Alot of this is just trying to acheive parity with some of the other CV module tests (this is a consistent style issue in almost all of them). Perhaps one future task is to refactor tests in all modules.

File: tests/test_blurry_classifier.py (hunk 5 above)
Comment (jgeaso1266) [reply tests/test_blurry_classifier.py:54]: Yes. Actually `fake_camera` does not implement the Camera interface because of the blurry attribute I added. Perhaps we can brainstorm together how to massage the fake camera to return a blurry image when we need it to.

File: src/blurry_classifier_module.py (hunk 14 above)
Comment (jgeaso1266) [reply src/blurry_classifier_module.py:180]: I changed it to actually convey what the number is.

File: Makefile (hunk 11 above)
Comment (jgeaso1266) [reply Makefile:17]: result from `make lint`:

``
src/blurry_classifier_module.py:182:28: E1101: Module 'cv2' has no 'Laplacian' member (no-member)
src/blurry_classifier_module.py:182:46: E1101: Module 'cv2' has no 'CV_64F' member (no-member)
``

File: src/blurry_classifier_module.py (hunk 13 above)
Comment (jgeaso1266) [reply src/blurry_classifier_module.py:184]: Let's discuss later today

File: Makefile (hunk 11 above)
Comment (penguinland) [reply Makefile:22]: I have misunderstood what E1101 is for. Put in a comment explaining why we're disabling these checks.

File: src/blurry_classifier_module.py (hunk 12 above)
Comment (penguinland) [reply src/blurry_classifier_module.py]: Strings you can just break up into shorter lines:
```python
raise ValueError(f"Camera name {camera_name} does not "
                 f"match the name {self.camera_name} "
                 f"in the config")
```

**Review by penguinland — APPROVED** [review]
LGTM! All my remaining feedback is easy to resolve (though some of it is still pretty important).

File: build.sh [hunk 18]
```diff
@@ -3,6 +3,8 @@ cd `dirname $0`
 
 . venv/bin/activate
 
+## using `pyrhon -m PyInastaller` as opposed to `pyinstaller` as the former
```
Comment (penguinland) [inline build.sh:6]: :laughing: `pyrhon -m PyInastaller` should be fixed next PR

File: src/main.py [hunk 19]
```diff
@@ -1,30 +1,7 @@
...
-    # )
-    module = Module.from_args()
-
-    module.add_model_from_registry(Vision.API, BlurryClassifier.MODEL)
-    await module.start()
-
 
 if __name__ == "__main__":
-    asyncio.run(main())
+    asyncio.run(Module.run_from_registry())
```
Comment (penguinland) [inline src/main.py:7]: Dang, this is way nicer than it used to be! Good idea
